src/v2/coins/eth.ts

```diff
@@ -805,6 +812,182 @@ class Eth extends BaseCoin {
+      };
+    }).call(this).asCallback(callback);
+  }
+
+  /**
+   * Validates that the hop prebuild from the HSM is valid and correct
+   * @param wallet The wallet that the prebuild is for
+   * @param hopPrebuild The prebuild to validate
+   * @param originalParams The original parameters passed to prebuildTransaction
+   * @param callback
+   * @returns void
+   * @throws Error if The prebuild is invalid
+   */
+  validateHopPrebuild(wallet: any, hopPrebuild: HopPrebuild, originalParams: any, callback: any): void {
+    return co(function *() {
+      const {
+        tx,
+        id,
+        signature,
+      } = hopPrebuild;
+
+      const {
+        recipients,
+      } = originalParams;
+
+      // first, validate the HSM signature
+      const bitgoKeychain: { pub: string } = yield this.keychains().get({ id: wallet.keyIds()[2] });
+      const bitgoPubkeyBuffer: Buffer = prova.HDNode.fromBase58(bitgoKeychain.pub).getPublicKeyBuffer();
+      const signatureBuffer: Buffer = Buffer.from(optionalDeps.ethUtil.stripHexPrefix(signature), 'hex');
+      const messageBuffer: Buffer = Buffer.from(optionalDeps.ethUtil.stripHexPrefix(id), 'hex');
+
+      const isValidSignature: boolean = secp256k1.verify(messageBuffer, signatureBuffer.slice(1), bitgoPubkeyBuffer);
+      if (!isValidSignature) {
+        throw new Error(`Hop txid signature invalid`);
+      }
+
+      // Then validate that the tx params actually equal the requested params
+      const originalAmount: string = recipients[0].amount;
+      const originalDestination: string = recipients[0].address;
+
+      const builtHopTx = new optionalDeps.EthTx(tx);
+      const hopAmount: string = BigNumber(optionalDeps.ethUtil.bufferToHex(builtHopTx.value)).toFixed(0);
+      const hopDestination: string = optionalDeps.ethUtil.bufferToHex(builtHopTx.to);
+      if (hopAmount !== originalAmount) {
+        throw new Error(`Hop amount: ${hopAmount} does not equal original amount: ${originalAmount}`);
+      }
+      if (hopDestination !== originalDestination) {
+        throw new Error(`Hop destination: ${hopDestination} does not equal original recipient: ${hopDestination}`);
+      }
+      if (!builtHopTx.verifySignature()) {
+        // We dont want to continue at all in this case, at risk of ETH being stuck on the hop address
+        throw new Error(`Invalid hop transaction signature, txid: ${id}`);
+      }
+      if (optionalDeps.ethUtil.addHexPrefix(builtHopTx.hash().toString('hex')) !== id) {
+        throw new Error(`Signed hop txid does not equal actual txid`);
+      }
+    }).call(this).asCallback(callback);
+  }
+
+  /**
+   * Gets the hop digest for the user to sign. This is validated in the HSM to prove that the user requested this tx
+   * @param paramsArr The parameters to hash together for the digest
+   */
+  static getHopDigest(paramsArr): Buffer {
+    const hash = new keccak('keccak256');
+    hash.update([Eth.hopTransactionSalt, ...paramsArr].join('$'));
```

Comment:
Hm, not sure how useful a public, static salt is here. What's the goal?

Reply:
@dpc's idea, its mostly to make sure that we dont ever accidentally (though super unlikely) signature collision by accident. Having the salt be super specifically for hop transactions, one has to deliberately be requesting this for the HSM to ever sign it. Signing [to, amount, gasprice, gaslimit, paymentId] isn't entirely implausible under a different circumstance